src/token-processor/images/image-cache.ts

```diff
@@ -0,0 +1,222 @@
+import { ENV } from '../../env';
+import { parseDataUrl, getFetchableDecentralizedStorageUrl } from '../util/metadata-helpers';
+import { logger } from '@hirosystems/api-toolkit';
+import { PgStore } from '../../pg/pg-store';
+import { Readable } from 'node:stream';
+import * as sharp from 'sharp';
+import * as fs from 'fs';
+import { Agent, fetch, request, errors } from 'undici';
+import {
+  ImageSizeExceededError,
+  ImageTimeoutError,
+  TooManyRequestsHttpError,
+  UndiciCauseTypeError,
+  ImageHttpError,
+  ImageParseError,
+} from '../util/errors';
+import { pipeline } from 'node:stream/promises';
+
+let gcsAuthToken: string | undefined;
+async function getGcsAuthToken(): Promise<string> {
+  if (gcsAuthToken !== undefined) return gcsAuthToken;
+  try {
+    const response = await request(
+      'http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token',
+      {
+        method: 'GET',
+        headers: { 'Metadata-Flavor': 'Google' },
+        throwOnError: true,
+      }
+    );
+    const json = (await response.body.json()) as { access_token: string };
+    // Cache the token so we can reuse it for other images.
+    gcsAuthToken = json.access_token;
+    return json.access_token;
+  } catch (error) {
+    throw new Error(`GCS access token error: ${error}`);
+  }
+}
+
+async function uploadImage(localPath: string, remoteName: string): Promise<string> {
```

Comment:
Could be a good time to switch to `https://www.npmjs.com/package/@google-cloud/storage` https://github.com/hirosystems/token-metadata-api/issues/215 it would simplify a lot of the auth and streaming here. Just a suggestion, this code seems fine too

Reply:
Yeah for sure... I'll prioritize that issue to work on it next